src/System.Private.CoreLib/shared/System/Number.Formatting.cs

```diff
@@ -2285,5 +2285,72 @@ private static uint Int64DivMod1E9(ref ulong value)
             value /= 1000000000;
             return rem;
         }
+
+        private static unsafe void DoubleToNumber(double value, int precision, ref NumberBuffer number)
+        {
+            number.precision = precision;
+
+            if (!double.IsFinite(value))
+            {
+                number.scale = double.IsNaN(value) ? ScaleNAN : ScaleINF;
+                number.sign = double.IsNegative(value);
+                number.digits[0] = '\0';
+            }
+            else if (value == 0.0)
+            {
+                number.scale = 0;
+                number.sign = double.IsNegative(value);
+                number.digits[0] = '\0';
+            }
+            else if (!Grisu3.Run(value, precision, ref number))
+            {
+                Dragon4(value, precision, ref number);
+            }
+        }
+
+        private static long ExtractFractionAndBiasedExponent(double value, out int exponent)
+        {
+            long fraction = GetMantissa(value);
```

Comment:
You are likely losing some cycles by converting double to long twice in a row.

Reply:
Fixed. 
-- I wish we had the ability to specify a function was `const` so that the JIT could just do the right/expected thing here.